### PR TITLE
Igor:Igorstats

### PR DESCRIPTION
**Pull request by aherna**
Adding Igor Statistics to master

### CODE REVIEW

**Comment by jcrussell** [comment]
Want to write some test cases?

**Comment by aherna** [comment]
> 
> 
> Want to write some test cases?

is this what you are looking for? 
[igor.log](https://github.com/sandia-minimega/minimega/files/2452044/igor.log)

**Comment by jcrussell** [comment]
I meant using go [testing](https://golang.org/pkg/testing/).

File: src/igor/stats.go [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+			return i, false
+		}
+	}
+	return -1, true
+}
+
+func (res *ResData) String() string {
+	var s string
+	var formatLong string = "2006-Jan-2-15:04"
+	s += "Reservation Name: " + res.ResName + "\tReservation ID: " + strconv.Itoa(res.ResId) + "\n"
```
[the remark this answers is not in the document]
Comment (aherna) [reply src/igor/stats.go]: changed to string builder per go documentation
https://golang.org/pkg/bytes/#Buffer.String

File: src/igor/stats.go [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+	TotalEarlyCancels    int
+	TotalExtensions      int
+	Reservations         map[string][]*ResData
+	Reservationsid       map[int]*ResData
+}
+
+//Main Stats function to output reservation calculations
+func runStats(_ *Command, _ []string) {
+	globalStats := Stats{}
+	globalStats.Reservations = make(map[string][]*ResData)
```
[the remark this answers is not in the document]
Comment (aherna) [reply src/igor/stats.go]: corrected

File: src/igor/stats.go [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+
+func init() {
+	// break init cycle
+	cmdStats.Run = runStats
+	cmdStats.Flag.StringVar(&statsD, "d", "", "")
+	cmdStats.Flag.BoolVar(&statsV, "v", false, "")
+
+}
+
+//Data type for individual Reservations
```
[the remark this answers is not in the document]
Comment (aherna) [reply src/igor/stats.go]: corrected

File: src/igor/stats.go [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+		if d > 0 {
+			stats.NumNodes += 1
+		}
+	}
+
+}
+
+//Returns Reservation data pointer. tries to find the reservation by unique ID
+//Otherwise it will search by name and by user
+func (stats *Stats) findRes(ru string, rn string, ri int, rs time.Time) (*ResData, bool) {
```
[the remark this answers is not in the document]
Comment (aherna) [reply src/igor/stats.go]: corrected

File: src/igor/stats.go [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+	if res, found := stats.Reservationsid[ri]; found {
+		return res, true
+	} else {
```
[the remark this answers is not in the document]
Comment (aherna) [reply src/igor/stats.go]: corrected

File: src/igor/stats.go [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+		case "INSTALL":
+			if !rs.After(time.Now().AddDate(0, 0, 0)) {
+				stats.addReservation(rn, ru, ri, rs, re, nodes)
+			}
+			break
+		case "DELETED":
+			ad, er := time.Parse(formatDateStamp, fields[DATESTAMP])
+			if er != nil {
+				log.Fatal("%v", er)
+				break
```
[the remark this answers is not in the document]
Comment (aherna) [reply src/igor/stats.go]: corrected

File: src/igor/stats.go [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+			}
+			at, er := time.Parse(formatTimeStamp, fields[TIMESTAMP])
+			if er != nil {
+				log.Fatal("%v", er)
+				break
```
[the remark this answers is not in the document]
Comment (aherna) [reply src/igor/stats.go]: corrected

File: src/igor/stats.go [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+	TotalDurationMinutes time.Duration
+	TotalEarlyCancels    int
+	TotalExtensions      int
+	Reservations         map[string][]*ResData
+	Reservationsid       map[int]*ResData
+}
+
+//Main Stats function to output reservation calculations
+func runStats(_ *Command, _ []string) {
+	globalStats := Stats{}
```
[the remark this answers is not in the document]
Comment (aherna) [reply src/igor/stats.go]: corrected

**Pull request merged**